File: elixir/phone-number/phone_number.exs
defmodule Phone do

  @number_length 10
  @area_code_range 0..2
  @prefix_range 3..5
  @line_number_range 6..9

  def number(formatted_number) do
    formatted_number
      |> strip_formatting
      |> validate
  end

  def area_code(number) do
    segment(number, @area_code_range)
  end

  def pretty(formatted_number) do
    n = number(formatted_number)
    "(#{area_code(n)}) #{prefix(n)}-#{line_number(n)}"
  end

  defp strip_formatting(formatted_number) do
    String.replace(formatted_number, %r/\D/, "")
  end

  defp validate(<< ?1, number :: [binary, size(@number_length)]>>), do: number
  defp validate(<< number :: [binary, size(@number_length)]>>), do: number
  defp validate(_), do: "0000000000"

  defp prefix(number) do
    segment(number, @prefix_range)
  end

  defp line_number(number) do
    segment(number, @line_number_range)
  end

  defp segment(number, range) do
    String.slice(number, Enum.at(range, 0), Enum.count(range))
  end

end

File: javascript/phone-number/phone-number.js
module.exports = function Phone(formattedNumber) {

  var digits = expectedDigits(withoutFormatting(formattedNumber));

  return {

    number: function number() {
      return digits;
    },

    areaCode: function areaCode() {
      return digits.substr(0, 3);
    },

    prefix: function prefix() {
      return digits.substr(3, 3);
    },

    lineNumber: function lineNumber() {
      return digits.substr(6, 4);
    },

    toString: function toString() {
      return '(' + this.areaCode() + ') ' + this.prefix() + '-' + this.lineNumber();
    }

  };

};

var US_PHONE_NUMBER_LENGTH = 10;
var US_COUNTRY_CODE = "1";
var INVALID_NUMBER = "0000000000";

function withoutFormatting(formattedNumber) {
  return formattedNumber.replace(/\D/g, '');
}

function expectedDigits(allDigits) {
  var countryCodeLength = US_COUNTRY_CODE.length;
  if(allDigits.length === US_PHONE_NUMBER_LENGTH) {
    return allDigits;
  }
  if(allDigits.length === US_PHONE_NUMBER_LENGTH + countryCodeLength &&
    allDigits.slice(0, countryCodeLength) === US_COUNTRY_CODE) {
    return allDigits.slice(countryCodeLength);
  }
  return INVALID_NUMBER;
}
